File: macos/Classes/CameraCapabilities.swift
import AVFoundation
import Foundation

public struct CameraCapabilities {
    public let supportsFocusPoint: Bool
    public let focusModes: [String]
    public let supportsExposurePoint: Bool
    public let exposureModes: [String]
    public let whiteBalanceModes: [String]
    public let supportsManualWhiteBalanceGains: Bool
    public let supportsManualExposure: Bool
    public let supportsHardwareZoom: Bool
    public let supportsTorch: Bool
    public let supportsMirroring: Bool
    public let supportsRotation: Bool
    public let supportedResolutions: [[String: Int]]
    public let supportedFormats: [[String: Any]]

    public var toMap: [String: Any] {
        return [
            "supportsFocusPoint": supportsFocusPoint,
            "focusModes": focusModes,
            "supportsExposurePoint": supportsExposurePoint,
            "exposureModes": exposureModes,
            "whiteBalanceModes": whiteBalanceModes,
            "supportsManualWhiteBalanceGains": supportsManualWhiteBalanceGains,
            "supportsManualExposure": supportsManualExposure,
            "supportsHardwareZoom": supportsHardwareZoom,
            "supportsTorch": supportsTorch,
            "supportsMirroring": supportsMirroring,
            "supportsRotation": supportsRotation,
            "supportedResolutions": supportedResolutions,
            "supportedFormats": supportedFormats,
        ]
    }
}

public enum CapabilityDiscovery {
    public static func discover(device: AVCaptureDevice,
                                 captureSession: AVCaptureSession?,
                                 previewLayer: AVCaptureVideoPreviewLayer?) -> CameraCapabilities {
        // Focus
        var focusModes: [String] = []
        if device.isFocusModeSupported(.autoFocus) { focusModes.append("auto") }
        if device.isFocusModeSupported(.continuousAutoFocus) { focusModes.append("continuous") }
        if device.isFocusModeSupported(.locked) { focusModes.append("locked") }
        let supportsFocusPoint = device.isFocusPointOfInterestSupported

        // Exposure
        var exposureModes: [String] = []
        if device.isExposureModeSupported(.autoExpose) { exposureModes.append("auto") }
        if device.isExposureModeSupported(.continuousAutoExposure) { exposureModes.append("continuous") }
        if device.isExposureModeSupported(.locked) { exposureModes.append("locked") }
        let supportsExposurePoint = device.isExposurePointOfInterestSupported

        // White balance (modes only)
        var whiteBalanceModes: [String] = []
        if device.isWhiteBalanceModeSupported(.autoWhiteBalance) { whiteBalanceModes.append("auto") }
        if device.isWhiteBalanceModeSupported(.continuousAutoWhiteBalance) { whiteBalanceModes.append("continuous") }
        if device.isWhiteBalanceModeSupported(.locked) { whiteBalanceModes.append("locked") }

        // Manual controls unavailable on macOS native
        let supportsManualWhiteBalanceGains = false
        let supportsManualExposure = false
        let supportsHardwareZoom = false

        // Torch
        let supportsTorch = device.hasTorch && (device.isTorchModeSupported(.on) || device.isTorchModeSupported(.auto))

        // Connection-based
        var supportsMirroring = false
        var supportsRotation = false
        if let captureSession = captureSession {
            for output in captureSession.outputs {
                if let connection = output.connection(with: .video) {
                    if connection.isVideoMirroringSupported { supportsMirroring = true }
                    if #available(macOS 14.0, *) {
                        if connection.isVideoRotationAngleSupported(90) { supportsRotation = true }
                    }
                }
            }
        }
        if !supportsMirroring, let connection = previewLayer?.connection {
            if connection.isVideoMirroringSupported { supportsMirroring = true }
            if #available(macOS 14.0, *) {
                if connection.isVideoRotationAngleSupported(90) { supportsRotation = true }
            }
        }

        // Resolutions (unique width x height) and collect distinct fps ranges without merging across formats
        struct FpsKey: Hashable { let w: Int; let h: Int; let min: Double; let max: Double }
        var resolutionsSet = Set<String>()
        var resolutions: [[String: Int]] = []
        var seenFormatKeys = Set<FpsKey>()
        var formats: [[String: Any]] = []
        for format in device.formats {
            let desc = format.formatDescription
            let dim = CMVideoFormatDescriptionGetDimensions(desc)
            let w = Int(dim.width), h = Int(dim.height)
            let resKey = "\(w)x\(h)"
            if !resolutionsSet.contains(resKey) {
                resolutionsSet.insert(resKey)
                resolutions.append(["width": w, "height": h])
            }
            for range in format.videoSupportedFrameRateRanges {
                // Keep original fractional values
                let minF = max(0.1, range.minFrameRate)
                let maxF = max(minF, range.maxFrameRate)
                let key = FpsKey(w: w, h: h, min: minF, max: maxF)
                if seenFormatKeys.contains(key) { continue }
                seenFormatKeys.insert(key)
                formats.append([
                    "width": w,
                    "height": h,
                    "minFps": minF,
                    "maxFps": maxF,
                ])
            }
        }
        // Stable sort formats by area then minFps then maxFps
        formats.sort { a, b in
            let aw = a["width"] as? Int ?? 0, ah = a["height"] as? Int ?? 0
            let bw = b["width"] as? Int ?? 0, bh = b["height"] as? Int ?? 0
            let aArea = aw * ah, bArea = bw * bh
            if aArea == bArea {
                let amin = (a["minFps"] as? Double) ?? 0
                let bmin = (b["minFps"] as? Double) ?? 0
                if amin == bmin {
                    let amax = (a["maxFps"] as? Double) ?? 0
                    let bmax = (b["maxFps"] as? Double) ?? 0
                    if amax == bmax { return aw < bw }
                    return amax < bmax
                }
                return amin < bmin
            }
            return aArea < bArea
        }

        return CameraCapabilities(
            supportsFocusPoint: supportsFocusPoint,
            focusModes: focusModes,
            supportsExposurePoint: supportsExposurePoint,
            exposureModes: exposureModes,
            whiteBalanceModes: whiteBalanceModes,
            supportsManualWhiteBalanceGains: supportsManualWhiteBalanceGains,
            supportsManualExposure: supportsManualExposure,
            supportsHardwareZoom: supportsHardwareZoom,
            supportsTorch: supportsTorch,
            supportsMirroring: supportsMirroring,
            supportsRotation: supportsRotation,
            supportedResolutions: resolutions,
            supportedFormats: formats
        )
    }
}

File: lib/camera_capabilities.dart
class CameraCapabilities {
  final bool supportsFocusPoint;
  final List<FocusMode> focusModes;
  final bool supportsExposurePoint;
  final List<ExposureMode> exposureModes;
  final List<WhiteBalanceMode> whiteBalanceModes;
  final bool supportsManualWhiteBalanceGains;
  final bool supportsManualExposure;
  final bool supportsHardwareZoom;
  final bool supportsTorch;
  final bool supportsMirroring;
  final bool supportsRotation;
  final List<Resolution> supportedResolutions;

  const CameraCapabilities({
    required this.supportsFocusPoint,
    required this.focusModes,
    required this.supportsExposurePoint,
    required this.exposureModes,
    required this.whiteBalanceModes,
    required this.supportsManualWhiteBalanceGains,
    required this.supportsManualExposure,
    required this.supportsHardwareZoom,
    required this.supportsTorch,
    required this.supportsMirroring,
    required this.supportsRotation,
    required this.supportedResolutions,
  });

  factory CameraCapabilities.fromMap(Map<String, dynamic> map) {
    List<String> _strings(dynamic v) =>
        (v as List?)?.cast<String>() ?? const [];
    List<Map<String, dynamic>> _mapList(dynamic v) =>
        (v as List?)
            ?.map((e) => Map<String, dynamic>.from(e as Map))
            .cast<Map<String, dynamic>>()
            .toList() ??
        const [];

    return CameraCapabilities(
      supportsFocusPoint: map['supportsFocusPoint'] == true,
      focusModes: _strings(map['focusModes'])
          .map(FocusModeX.fromString)
          .whereType<FocusMode>()
          .toList(),
      supportsExposurePoint: map['supportsExposurePoint'] == true,
      exposureModes: _strings(map['exposureModes'])
          .map(ExposureModeX.fromString)
          .whereType<ExposureMode>()
          .toList(),
      whiteBalanceModes: _strings(map['whiteBalanceModes'])
          .map(WhiteBalanceModeX.fromString)
          .whereType<WhiteBalanceMode>()
          .toList(),
      supportsManualWhiteBalanceGains:
          map['supportsManualWhiteBalanceGains'] == true,
      supportsManualExposure: map['supportsManualExposure'] == true,
      supportsHardwareZoom: map['supportsHardwareZoom'] == true,
      supportsTorch: map['supportsTorch'] == true,
      supportsMirroring: map['supportsMirroring'] == true,
      supportsRotation: map['supportsRotation'] == true,
      supportedResolutions: _mapList(map['supportedResolutions'])
          .map((e) => Resolution(
                width: (e['width'] as num?)?.toInt() ?? 0,
                height: (e['height'] as num?)?.toInt() ?? 0,
              ))
          .toList(),
    );
  }
}

class Resolution {
  final int width;
  final int height;
  const Resolution({required this.width, required this.height});
  @override
  String toString() => '${width}x$height';
}

enum FocusMode { auto, continuous, locked }

class FocusModeX {
  static String name(FocusMode m) {
    switch (m) {
      case FocusMode.auto:
        return 'auto';
      case FocusMode.continuous:
        return 'continuous';
      case FocusMode.locked:
        return 'locked';
    }
  }

  static FocusMode? fromString(String s) {
    switch (s) {
      case 'auto':
        return FocusMode.auto;
      case 'continuous':
        return FocusMode.continuous;
      case 'locked':
        return FocusMode.locked;
    }
    return null;
  }
}

enum ExposureMode { auto, continuous, locked }

class ExposureModeX {
  static String name(ExposureMode m) {
    switch (m) {
      case ExposureMode.auto:
        return 'auto';
      case ExposureMode.continuous:
        return 'continuous';
      case ExposureMode.locked:
        return 'locked';
    }
  }

  static ExposureMode? fromString(String s) {
    switch (s) {
      case 'auto':
        return ExposureMode.auto;
      case 'continuous':
        return ExposureMode.continuous;
      case 'locked':
        return ExposureMode.locked;
    }
    return null;
  }
}

enum WhiteBalanceMode { auto, continuous, locked }

class WhiteBalanceModeX {
  static String name(WhiteBalanceMode m) {
    switch (m) {
      case WhiteBalanceMode.auto:
        return 'auto';
      case WhiteBalanceMode.continuous:
        return 'continuous';
      case WhiteBalanceMode.locked:
        return 'locked';
    }
  }

  static WhiteBalanceMode? fromString(String s) {
    switch (s) {
      case 'auto':
        return WhiteBalanceMode.auto;
      case 'continuous':
        return WhiteBalanceMode.continuous;
      case 'locked':
        return WhiteBalanceMode.locked;
    }
    return null;
  }
}
